### ISSUE TITLE
[cpp] Abstract class functions with default values are broken

### ISSUE THREAD

**Issue by Aidan63** [HaxeFoundation/haxe#11666]
Calling a function on an abstract class which has default values has seemingly never worked properly, I just spotted this as it generates invalid C++ when compiled with -D sciptable, but upon closer inspection they don't actually worked for non cppia enabled compiles either.

```haxe
abstract class Foo {
    abstract public function foo(v:Int = 0):Void;
}

class ConcreteFoo extends Foo {
    public function new() {}
    
    public function foo(v:Int = 0) {
        trace(v);
    }
}

function main() {
    final o = new ConcreteFoo();

    o.foo(7);
    
    (o : Foo).foo(12);
}
```

The way they're broken is a bit hard to catch, in the above example you'd expect 7 then 12 to be traced, but you only see 7 and then the program exits (no crash / exception, a graceful exit). Invoking these functions on the concrete type is fine, but trying to use the abstract base is where the issues come into play.
The generated C++ class from the abstract class generates `virtual foo(int v) {}` where as the concrete class gets the correct `virtual foo(::hx::Null<int> __o_v)`, but since these are different signatures they're different functions, so calling `foo` on the base invokes that empty function...hence no crash.

Issue seems to be the abstract class path calls `ctx_arg` with no default expression so it doesn't think the abstract version has a default value, I've got a WIP fix which checks for the `value` meta which seems to be where abstract classes store default argument expressions.

**Comment by tobil4sk** [comment]
> The way they're broken is a bit hard to catch, in the above example you'd expect 7 then 12 to be traced, but you only see 7 and then the program exits (no crash / exception, a graceful exit)

Ran into this in the haxelib codebase, was a nightmare to figure out what was going wrong... 😅 It was working before as expected, but randomly stops working in a haxelib commit that doesn't even touch the broken method...

I wonder if we could add C++11's `override` to help prevent problems like this happening again. This way, the broken C++ code would fail to compile, and at least we don't generate a broken program. I guess it would require compatibility wrappers for pre c++11.

**Comment by Aidan63** [comment]
I think the best fix for this would be to get abstract functions implemented as pure virtual functions. They were originally done this way but I switched them over to those no-op virtual functions as cppia's scaffolding needs to construct the classes.

Override might help here as well, I did update the generator to add `override` where possible but that led to a lot of weird c++ errors I couldn't be bothered to track down and gave up.
